src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/Utilities/BaseTest.cs

```diff
@@ -151,25 +155,115 @@ public void AfterAllTests ()
 			}
 		}
 
-		protected void WaitFor(int milliseconds)
+		protected static void WaitFor(int milliseconds)
 		{
 			var pause = new ManualResetEvent(false);
 			pause.WaitOne(milliseconds);
 		}
 
-		protected static string RunAdbCommand (string command, bool ignoreErrors = true)
+		protected static string RunAdbCommand (string command, bool ignoreErrors = true, int timeout = 30)
 		{
 			string ext = Environment.OSVersion.Platform != PlatformID.Unix ? ".exe" : "";
 			string adb = Path.Combine (AndroidSdkPath, "platform-tools", "adb" + ext);
 			var proc = System.Diagnostics.Process.Start (new System.Diagnostics.ProcessStartInfo (adb, command) { RedirectStandardOutput = true, RedirectStandardError = true, UseShellExecute = false });
-			if (!proc.WaitForExit ((int)TimeSpan.FromSeconds (30).TotalMilliseconds)) {
+			if (!proc.WaitForExit ((int)TimeSpan.FromSeconds (timeout).TotalMilliseconds)) {
 				proc.Kill ();
 				proc.WaitForExit ();
 			}
 			var result = proc.StandardOutput.ReadToEnd ().Trim () + proc.StandardError.ReadToEnd ().Trim ();
 			return result;
 		}
 
+		protected static void RunAdbInput (string command, params object [] args)
+		{
+			RunAdbCommand ($"shell {command} {string.Join (" ", args)}");
+		}
+
+		protected static string ClearAdbLogcat ()
+		{
+			return RunAdbCommand ("logcat -c");
+		}
+
+		protected static string GetAdbLogcat (int timeout = 1)
+		{
+			return RunAdbCommand ("logcat", timeout: timeout);
+		}
+
+		protected static void AdbStartActivity (string activity)
+		{
+			RunAdbCommand ($"shell am start -S -n \"{activity}\"");
+		}
+
+		protected static bool MonitorAdbLogcat (Func<string, bool> action, int timeout = 10)
+		{
+			string ext = Environment.OSVersion.Platform != PlatformID.Unix ? ".exe" : "";
+			string adb = Path.Combine (AndroidSdkPath, "platform-tools", "adb" + ext);
+			var proc = Process.Start (new ProcessStartInfo (adb, "logcat") { RedirectStandardOutput = true, RedirectStandardError = true, UseShellExecute = false });
+			var sw = new Stopwatch ();
+			try {
+				TimeSpan time = TimeSpan.FromSeconds (timeout);
+				while (time.TotalMilliseconds > 0) {
+					sw.Start ();
+					if (action (proc.StandardOutput.ReadLine ()))
+						return true;
+					time = time.Subtract (TimeSpan.FromMilliseconds (sw.ElapsedMilliseconds));
+					sw.Reset ();
+				}
+			} finally {
+				sw.Stop ();
+				proc.Kill ();
+				proc.WaitForExit ();
+			}
+			return false;
+		}
```

Comment:
Since this adds another 100 lines to `BaseTest` should we consider moving all adb-related stuff to another class? `AdbTest : BaseTest` might be the easiest.

Then all our tests won't inherit this adb-stuff most don't use right now.